From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tavian Barnes <tavianator@tavianator.com>
Date: Fri, 18 Nov 2022 09:52:21 -0500
Subject: [PATCH] 4759: decompiler: Don't propagate types through call clobbers

INDIRECT ops that model call-clobbered registers indicate that the value
may be changed arbitrarily by the call, so it doesn't make sense to
propagate types through them to the value it had before the call.

Fixes: #4639

Sponsored-by: RCS Lab at University of Waterloo, NSERC, WHJIL
---
 Ghidra/Features/Decompiler/src/decompile/cpp/heritage.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/heritage.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/heritage.cc
index 73ce97c0c2..06320721b2 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/heritage.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/heritage.cc
@@ -1510,6 +1510,7 @@ void Heritage::guardCalls(uint4 fl,const Address &addr,int4 size,vector<Varnode
     // We do not guard the call if the effect is "unaffected" or "reload"
     if ((effecttype == EffectRecord::unknown_effect)||(effecttype == EffectRecord::return_address)) {
       indop = fd->newIndirectOp(fc->getOp(),addr,size,0);
+      indop->setStopTypePropagation();
       indop->getIn(0)->setActiveHeritage();
       indop->getOut()->setActiveHeritage();
       write.push_back(indop->getOut());
@@ -1520,6 +1521,7 @@ void Heritage::guardCalls(uint4 fl,const Address &addr,int4 size,vector<Varnode
     }
     else if (effecttype == EffectRecord::killedbycall) {
       indop = fd->newIndirectCreation(fc->getOp(),addr,size,possibleoutput);
+      indop->setStopTypePropagation();
       indop->getOut()->setActiveHeritage();
       write.push_back(indop->getOut());
     }
-- 
2.45.1

